File: Learning/ls01.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

public class learn05 {
    public static void main(String args[] ) throws Exception {

		Island	b = new Island();
		b.setBoard();	//b.dspBoard();
		b.labeling();	//b.dspBoard();
		b.catBoard();	//b.dspBoard();
		b.getBoard();
    }
}

class Island {

	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	int		col;
	int		row;
	int[][]	board;

	Island() throws Exception {
		String s = br.readLine();
        s = s.trim();
        String[] sa = s.split(" ");
		col = Integer.parseInt(sa[0]);
		row = Integer.parseInt(sa[1]);
        board = new int[row][col];
	}

    void dspBoard() {
		System.out.println("row:"+row);
		System.out.println("col:"+col);
		for (int i=0; i<row; i++) {
			for (int j=0; j<col; j++) {
                System.out.print(" "+board[i][j]);
            }
			System.out.println();
        }
    }

    void setBoard() throws Exception {
		for (int i=0; i<row; i++) {
			String s = br.readLine();
        	s = s.trim();
        	String[] sa = s.split(" ");
			for (int j=0; j<col; j++) {
                board[i][j] = Integer.parseInt(sa[j]);
            }
        }
    }

    void labeling() {
        int label = 2;
		ArrayList<Integer> list = new ArrayList<Integer>();
		int[][] arrays = {{0,-1},{-1,0}};
        int x,y;
		for (int i=0; i<row; i++) {
			for (int j=0; j<col; j++) {
                if (board[i][j] == 1) {
                    list.clear();
					for (int[] pos : arrays ) {
						x = pos[0];
						y = pos[1];
						if ( (i+y)<0 || (j+x)<0 )
							continue;
                    	if (board[i+y][j+x] > 1) {
                    		list.add( board[i+y][j+x] );
                    	}
                	}
					if (list.size() == 0) {
                        board[i][j] = label;
                        label += 1;
                    } else {
                        board[i][j] = Collections.min(list);
                	}
            	}
        	}
    	}
    }

    boolean cat_x() {
        boolean flg = false;
		ArrayList<Integer> list = new ArrayList<Integer>();
        int k,min;
		for (int i=0; i<row; i++) {
			for (int j=0; j<col; j++) {
                // ‰¡•ûŒü‚Ì˜AŒ‹
                if (board[i][j] > 1) {
                    list.clear();
                    k = j;
                    while (k < col && board[i][k] > 1) {
						list.add( board[i][k] );
                        k += 1;
                    }
                    k = j;
                    min = Collections.min(list);
                    while (k < col && board[i][k] > 1) {
                        if (board[i][k] != min) {
                            board[i][k] = min;
                            flg = true;
                        }
                        k += 1;
                    }
                    if (k < col ) {
                        j = k;
                    } else {
                        j = col;
                    }
                }
            }
        }
        return flg;
    }
    
    boolean cat_y() {
        boolean flg = false;
		ArrayList<Integer> list = new ArrayList<Integer>();
        int k,min;
		for (int j=0; j<col; j++) {
			for (int i=0; i<row; i++) {
                // c•ûŒü‚Ì˜AŒ‹
                if (board[i][j] > 1) {
                    list.clear();
                    k = i;
                    while (k < row && board[k][j] > 1) {
						list.add( board[k][j] );
                        k += 1;
                    }
                    k = i;
                    min = Collections.min(list);
                    while (k < row && board[k][j] > 1) {
                        if (board[k][j] != min) {
							board[k][j] = min;
                            flg = true;
                        }
                        k += 1;
                    }
                    if (k < row) {
                        i = k;
                    } else {
                        i = row;
                    }
                }
            }
        }
        return flg;
    }

    void catBoard() {
        while (true) {
            boolean flg_x = cat_x();
            boolean flg_y = cat_y();
            if (flg_x == false && flg_y == false) {
                break;
            }
        }
    }

    void getBoard() {
		HashMap<Integer,Integer> hash = new HashMap<Integer,Integer>();
		for (int i=0; i<row; i++) {
			for (int j=0; j<col; j++) {
                if (board[i][j] != 0) {
                    if (!hash.containsKey(board[i][j])) {
                        hash.put(board[i][j],1);
                    }
                }
            }
        }
		System.out.println(hash.size());
    }
}

File: Learning/ls01.rb
# _*_ coding: cp932 _*_
class Island
    def initialize()
        s = gets.chomp.split(" ")
        @col =  s[0].to_i
        @row =  s[1].to_i
        @board = Array.new(@row).map{Array.new(@col,0)}
    end
    def setBoard()
        @row.times {|i|
            s = gets.chomp.split(" ")
            @col.times {|j|
                @board[i][j] = s[j].to_i
            }
        }
    end
    def labeling()
        label = 2
        @row.times {|i|
            @col.times {|j|
                if @board[i][j] == 1 then
                    list = Array.new
                    # 上      左横
                    [[0,-1], [-1,0]].each {|x,y|
                        #next if !(0..(@row-1)).include?(i+y) ||
                        #    !(0..(@col-1)).include?(j+x)
                        next if (i+y)<0 || (j+x)<0
                        if @board[i+y][j+x] > 1 then
                            list << @board[i+y][j+x]
                        end
                    }
                    if list.length == 0 then
                        @board[i][j] = label
                        label += 1
                    else
                        @board[i][j] = list.min
                    end
                end
            }
        }
    end
    def cat_x()
        flg = false
        @row.times {|i|
            @col.times {|j|
                # 横方向の連結
                if @board[i][j] > 1 then
                    list = Array.new
                    k = j
                    while k < @col && @board[i][k] > 1 do
                        list << @board[i][k]
                        k += 1
                    end
                    k = j
                    min = list.min
                    while k < @col && @board[i][k] > 1 do
                        if @board[i][k] != min then
                            @board[i][k] = min
                            flg = true
                        end
                        k += 1
                    end
                    if k < @col then
                        j = k
                    else
                        j = @col
                    end
                end
            }
        }
        return flg
    end
    def cat_y()
        flg = false
        @col.times {|j|
            @row.times {|i|
                # 縦方向の連結
                if @board[i][j] > 1 then
                    list = Array.new
                    k = i
                    while k < @row && @board[k][j] > 1 do
                        list << @board[k][j]
                        k += 1
                    end
                    k = i
                    min = list.min
                    while k < @row && @board[k][j] > 1 do
                        if @board[k][j] != min then
                            @board[k][j] = min
                            flg = true
                        end
                        k += 1
                    end
                    if k < @row then
                        i = k
                    else
                        i = @row
                    end
                end
            }
        }
        return flg
    end
    def catBoard()
        while true do
            flg_x = cat_x()
            flg_y = cat_y()
            if flg_x == false && flg_y == false then
                break
            end
        end
    end
    def getBoard()
        hash = Hash.new()
        @row.times {|i|
            @col.times {|j|
                if @board[i][j] != 0 then
                    if !hash.key?(@board[i][j]) then
                        hash[@board[i][j]] = 1
                    else
                        hash[@board[i][j]] += 1
                    end
                end
            }
        }
        puts hash.length
    end
end

b = Island.new()
b.setBoard()
b.labeling()
b.catBoard()
b.getBoard()
